### DISCUSSION TITLE
How to dispatch Action from File Menu to Java server without the Diagram Module

### DISCUSSION THREAD

**Discussion by amiths89** [eclipse-glsp/glsp#1332] [General] [answered]
I am tryin to dispatch an action from a Command Contribution created on the client side which basically creates a new resource file in the workspace. Now I have a handler on the Java server side but since this is not related to the diagram am not sure how to bind this Action handler.
I have a Menu contribution and command contribution bound in the front end that dispatches the action and then an Action and Action handler on the Java server side as well. The below code is the Command contribution. I need this to dispatch the action handler in the Java server. But it does not seem to be triggering the action.
`
constructor(@inject(TYPES.IActionDispatcher) protected actionDispatcher: GLSPActionDispatcher) {}
registerCommands(registry: CommandRegistry): void {
        registry.registerCommand(startUserProgramCommand, {
            execute: () => {
                this.actionDispatcher?.dispatch(CreatNewProjectOperation.create());
            }
        });
`

**Comment by planger** [comment]
I assume, you are talking about a Theia command contribution in the Theia integration, is that right?

If yes, I recommend registering the Theia command as you did, but use the [`GLSPCommandHandler`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/packages/theia-integration/src/browser/diagram/glsp-command-handler.ts#L62) instead of a direct handler (`execute: () => {...}`).

With the `GLSPCommandHandler`, you get access to the GLSP diagram editor `context` and just return GLSP actions to be invoked as a result of the Theia command. Please see the workflow diagram for [an example](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/examples/workflow-theia/src/browser/diagram/workflow-task-editing-context-menu.ts#L29).

```ts
@injectable()
export class WorkflowTaskEditCommandContribution implements CommandContribution {
    @inject(ApplicationShell) protected readonly shell: ApplicationShell;
    registerCommands(commands: CommandRegistry): void {
        commands.registerCommand(
            { id: WorkflowTaskEditingCommands.EDIT_TASK, label: 'Direct Edit Task' },
            new GLSPCommandHandler(this.shell, {
                actions: context => /* the array of actions to be invoked by this command */ [
                    SetUIExtensionVisibilityAction.create({
                        extensionId: TaskEditor.ID,
                        visible: true,
                        contextElementsId: [context.selectedElements[0].id]
                    })
                ],
                isEnabled: context => !context.isReadonly && context.selectedElements.filter(isTaskNode).length === 1
            })
        );
    }
}
```

If the server registers an action handler for the action, you invoke in your command, the action will be forwarded to the server automatically and your handler on the server will be executed accordingly.

**Reply by amiths89** [reply]
Thanks @planger . also i had some logic in the client side for user input and dialog. The reason I was using the execute was because of a dialog and userinput. Would that still be possible with GLSPCommandHandler?

     `registry.registerCommand(startUserProgramCommand, {
            execute: () => {
                const filename = 'untitled_project';
                const parentUri = this.workspaceService.workspace?.resource;
                if (parentUri) {
                    const targetUri = parentUri.resolve(filename);
                    const vacantChildUri = FileSystemUtils.generateUniqueResourceURI(this.workspaceService.workspace, targetUri, false);
                    const dialog = new WorkspaceInputDialog(
                        {
                            title: nls.localizeByDefault('New Project...'),
                            maxWidth: 400,
                            parentUri: parentUri,
                            initialValue: vacantChildUri.path.base,
                            placeholder: nls.localize('theia/workspace/newFilePlaceholder', 'File Name')
                        },
                        this.labelProvider
                    );
                    dialog.open().then(async name => {
                        if (name) {
                            const fileUri = parentUri.resolve(name);
                            await this.fileService.createFolder(fileUri);                            
                            this.workspaceService.close();
                            this.workspaceService.open(this.workspaceService.workspace?.resource as URI);
                            this.actionDispatcher?.dispatch(CreatNewProjectOperation.create());
                        }
                    });
                }
            }
        });`

**Reply by planger** [reply]
I think, you could override `execute()` of `GLSPCommandHandler` and first do your stuff as above and then call `super()`.
Alternatively, you could write a plain Theia command handler as you have above, which at the end triggers another command that is then handled by a `GLSPCommandHandler`, which just triggers a GLSP action.

**Reply by amiths89** [reply]
oh yes that should work. Thanks again @planger

**Reply by amiths89** [reply]
@planger I tried the alternate way by executing as part of separate command at the end of the execute() `registry.executeCommand(JavaServerCommand.id);` My Action still does not seem to register on the server side. I have an Action and AbstractActionHandler associated to the action on the Java Server side with same ID as the action ID associated to the JavaServerCommand. Also I had to override the isEnabled in the GLSPCommandHandler to make the handler enabled for executing the command. The command is getting executed but it does not seem to trigger the server side. Would there be something else needed to register on the server side

**Reply by planger** [reply]
No, I don't think you'd need to register something else. Just to make sure: You need to trigger another *Theia* command (not the GLSP action), which is handled by a `GLSPCommandHandler`, which then in turn triggers the GLSP Action that is handled in the Java GLSP server.

**Reply by amiths89** [reply]
Yea I think am doing that way. Is it like this?(I ve commented the logic for dialog for now)
`registry.registerCommand(
            JavaServerCommand,
            new GLSPCommandHandler(this.shell, {
                actions: context => [CreateNewProjectAction.create({ input: 'newProject' })]
            })
        );`
`registry.registerCommand(NewProjectCommand, {
            execute: () => {
                registry.executeCommand('JavaServer.Command');
            }
        });`

**Reply by planger** [reply]
Yes, I guess. At least for the part that I'm seeing in the extract of your code, it looks reasonable.

**Reply by amiths89** [reply]
So looking at GLSPCommandHandler it looks like it is expecting a diagram widget in the execute function.
`this.diagramWidget.actionDispatcher.dispatchAll(actions);`
But in my case I am triggering this function from the File Menu to start a new Project.

**Reply by planger** [reply]
Ah, this was the missing piece! The `GLSPCommandHandler` is designed to retrieve the currently active GLSP diagram widget from Theia's application shell. This is for use cases like commands and context menus in a GLSP widget. If you use that for application menus (as GLSP does it with the Diagram menu), the menu should only be visible if a GLSP widget is active. If there is no GLSP active, the `GLSPCommandHandler` can't know to which diagram editor it should dispatch actions to.

If you need an application menu that is executed outside the context of a specific GLSP diagram editor, you'll need to implement the command handler independently of a GLSP server until you are ready to open one of the generated resources in a GLSP editor (e.g., with Theia's `OpenerService`). Once the diagram is opened, you can trigger further actions in the GLSP editor, e.g. by passing arguments to the open call directly, or by again using the `GLSPCommandHandler` (which now should work as the GLSP editor is open).

**Reply by amiths89** [reply]
@planger Actually I wanted to look from outside GLSP diagram completely without touching the diagrams at all. This has to do with creating the diagram files(XML content) and other resource files so you see I dont want to open the diagram as am trying to populate the content of the diagram files before it being open. At the moment I am overriding the loadResource in the Java server to populate the contents of the diagram files which triggers when i try to open the diagram but this is only a temporary solution as I would need to work on more actions from the Theia Menu that would not need to touch the diagrams. I am currently trying to figure out the json rpc messages being sent to trigger the actions. What would i need to do if i wanted to trigger actions bypassing the diagrams completely?

**Reply by planger** [reply]
If you want to add commands and menu items in Theia that don't touch the diagram-related aspects, e.g. creating a new file with a certain content, you can just use plain [Theia commands](https://theia-ide.org/docs/commands_keybindings/). There is nothing specific to GLSP then, and thus no `GLSPCommandHandler` involved.

Only if you want to trigger a Theia command that impacts a running diagram instance you probably need `GLSPCommandHandler` for that.

**Reply by amiths89** [reply]
I understand this is not the way GLSP works but if i were to explore this I wanted a hint of a direction where I should be looking at in order to be in a position to run my Java code that is outside the diagram.

So my file editing code which is in Java are all sitting in the glsp-server, I have a Theia Command in glsp-client that promts the user to enter a project name which then creates a folder and an empty diagram file and then using that file path I need to be able run the Java code that is currently in the glsp-server which edits the contents of the file.

**Reply by tortmayr** [reply, marked as answer]
You should be able to make this work with a little bit of customization.

In order to communicate with the GLSP server you need to use the `GLSPClient`. If you want to retrieve the client ouside of the context of an active diagram you can use the `DiagramServiceProvider`.

So in your custom command contribution, you can inject the `DiagramServiceProvider` and retrieve the `GLSPClient` for your contributionId, then send the action message. Here is a snippet of a potential implementation:

```ts
@injectable()
export class MyCommandContribution implements CommandContribution {
    @inject(DiagramServiceProvider)
    protected diagramServiceProvider: DiagramServiceProvider;

    protected clientContribution: GLSPClientContribution;

    @postConstruct()
    protected init(): void {
        this.clientContribution = this.diagramServiceProvider.getGLSPClientContribution('yourContributionId');
    }

    registerCommands(commands: CommandRegistry): void {
        commands.registerCommand(startUserProgramCommand, {
            execute: async () => {
                const glspClient = await this.clientContribution.glspClient;
                glspClient.sendActionMessage({ clientId: 'no-session', action: CreateNewProjectAction.create() });
            }
        });
    }
}
```
Then on the server side you need a custom `GLSPServer` implementation, because the default implementation will faill if it receives an action that is not associated with a diagram id. For this you need to override the `process` method:

```java
public class MyServer extends DefaultGLSPServer {

   @Override
   public void process(final ActionMessage message) {
      validateServerInitialized();
      if (message.getClientId().equals("no-session")) {
         this.handleSessionLessAction(message.getAction());
         return;
      }
      super.process(message);
   }

   protected void handleSessionLessAction(final Action action) {
      if (action instanceof CreateNewProjectAction) {
         // your custom handling
      }
   }
}

```
and create a custom server module:
```java
public class MyServerModule extends ServerModule {

   @Override
   protected Class<? extends GLSPServer> bindGLSPServer() {
      return MyServer.class;
   }
}
```
and then use this new custom server module instead of the default one in your main/launch class.

I believe this approach should work, but I have not tested it.

**Reply by amiths89** [reply]
oh yes this is exactly what I was looking for. This will get me started. Thanks @tortmayr  and @planger.

**Reply by amiths89** [reply]
I tried this solution and it worked perfectly fine and am able to trigger actions without opening the diagrams. Additional to this solution I only had to bind it to the diagram in the server side I guess it was looking for the module associated to the contribution ID. Thanks again @planger & @tortmayr
